### PR TITLE
google oauth2 support using apache oltu

### PR DESCRIPTION
**Pull request by krro**

### CODE REVIEW

File: src/main/resources/application.conf [hunk 1]
```diff
@@ -0,0 +1,16 @@
+oauth {
+  client {
+    id = "93705780189-sb3je3l8bng3f9gd5popvg19g52rrq8r.apps.googleusercontent.com"
+    secret = "4E_xUGLWg8kk5NXrnPevAWQF"
```
Comment (krro) [inline src/main/resources/application.conf]: :( Forgot to remove those before commit :(

File: src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    redirectUrl = "http://arts.kainos.pl:8080"
```
Comment (krro) [inline src/main/resources/application.conf]: ofc, can be used with /etc/hosts for now

File: src/main/scala/com/kainos/rest/ArtsRoutes.scala [hunk 3]
```diff
@@ -1,22 +1,15 @@
...
-  val artsRepository = new ArtsRepository
+trait ArtsRoutes extends JsonProtocol {
 
-  val staticAssets: Route =
-    path("") {
-      getFromResource("web/index.html")
-    } ~
-    get {
-      getFromResourceDirectory("web")
-    }
```
Comment (krro) [inline src/main/scala/com/kainos/rest/ArtsRoutes.scala]: Some movement across various files

File: src/main/scala/com/kainos/rest/Rest.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  val staticAssets: Route =
+    path("") {
+      getFromResource("web/index.html")
+    } ~
+    get {
+      getFromResourceDirectory("web")
+    }
+
+  val myRoutes =
```
Comment (krro) [inline src/main/scala/com/kainos/rest/Rest.scala]: Btw. possibly this can be changed to sth like ```appRoutes```

File: build.gradle [hunk 5]
```diff
@@ -38,6 +38,9 @@ dependencies {
   compile group: 'com.typesafe', name: 'config', version: '1.3.0'
   compile group: 'org.mongodb.scala', name: 'mongo-scala-driver_2.11', version: '1.1.0'
   compile group: 'joda-time', name: 'joda-time', version: '2.9.3'
+  compile group: 'org.apache.oltu.oauth2', name: 'org.apache.oltu.oauth2.client', version:'1.0.1'
+  compile group: 'javax.servlet', name: 'javax.servlet-api', version:'3.1.0'
+  compile group: 'org.mockito', name: 'mockito-core', version:'1.10.19'
```
Comment (krro) [inline build.gradle]: 'compile' scope is not a mistake for mockito. this was done on purpose as I needed to mock sth later on in production code

File: src/main/resources/application.conf (hunk 1 above)
Comment (lukgad) [inline src/main/resources/application.conf]: Can we even store this in repo? Looks very unsecure :(

**Comment by lukgad** [comment]
Looks ok, but I would think about adding something to code to make it more descriptive. I know, good code doesn't need comments but we have to at least 2 goals here:
- promote scala and/or functional/reactive programing in asynchronous systems
- **learn**
- have fun
So I think we can add some comments/link to classes like SessionRoutes.

**Comment by lukgad** [comment]
Also, no unit tests so far.

File: src/main/scala/com/kainos/rest/SessionRoutes.scala [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    .setScope(scope)
+    .buildQueryMessage()
+
+  val sessionRoutes: Route =
+    path("login") {
+      redirect(authorizationRequest.getLocationUri, StatusCodes.SeeOther)
+    } ~ path("oauth2callback") {
+      parameterMultiMap { map =>
+
+        val httpRequest = Mockito.mock(classOf[HttpServletRequest])
```
Comment (lukgad) [inline src/main/scala/com/kainos/rest/SessionRoutes.scala]: Why is this needed? Can you explain?

File: src/main/scala/com/kainos/rest/SessionRoutes.scala [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+        val bearerClientRequest = new OAuthBearerClientRequest(userInfoEndpoint)
+          .setAccessToken(accessToken.getAccessToken)
+          .buildQueryMessage
+
+        val resourceResponse = oAuthClient.resource(bearerClientRequest, OAuth.HttpMethod.GET, classOf[OAuthResourceResponse])
+
+        println("token received")
+
+        if(provider == "google") {
```
Comment (lukgad) [inline src/main/scala/com/kainos/rest/SessionRoutes.scala]: Maybe enumeration of supported providers can be used here?

File: src/main/scala/com/kainos/rest/SessionRoutes.scala [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+        }
+
+        redirect("/", StatusCodes.SeeOther)
+      }
+    }
+
+  def userId(resourceResponse: OAuthResourceResponse) = {
+
+    case class GoogleUserInfoRequest(emailAddresses: Seq[GoogleEmailAddress])
```
Comment (lukgad) [inline src/main/scala/com/kainos/rest/SessionRoutes.scala]: This class is completly google-centric. I would extract this to trait and merge this with provider from line 76.

**Comment by lukgad** [comment]
Review, comments added.

File: src/main/scala/com/kainos/rest/SessionRoutes.scala (hunk 6 above)
Comment (krro) [inline src/main/scala/com/kainos/rest/SessionRoutes.scala]: I needed to call ```OAuthAuthzResponse.oauthCodeAuthzResponse(httpRequest)``` and this Oltu library assumes that it will get the parameters from object of type ```HttpServletRequest``` (as strange as this may sound :( ) I don't know if I can easily create an instance of that complex interface - using mockito for this purpose was the quickest solution that I could think of
